### PR TITLE
Cache more implicit scopes

### PR DESCRIPTION
**Pull request by odersky**
Also cache implicit scopes that arise from types that map to a different
lifted type.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
Shows an improvement of 5% for Dotty. I don't think the other full compilation benchmarks use implicits much, so I would not expect them to be much affected.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -314,20 +315,19 @@ object Implicits:
 
     /** The implicit references that are eligible for type `tp`. */
     def eligible(tp: Type): List[Candidate] =
-      if (tp.hash == NotCached) computeEligible(tp)
+      if (tp.hash == NotCached)
+        Stats.record(i"compute eligible not cached ${tp.getClass}")
+        Stats.record(i"compute eligible not cached")
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala:320]:
```suggestion
```

**Review by liufengyun — APPROVED** [review]
It's impressive, LGTM 👍

**Pull request merged**